src/main/java/org/apache/solr/benchmarks/task/UrlCommandTask.java

```diff
@@ -0,0 +1,152 @@
+package org.apache.solr.benchmarks.task;
+
+import io.prometheus.client.Histogram;
+import org.apache.solr.benchmarks.BenchmarkContext;
+import org.apache.solr.benchmarks.BenchmarksMain;
+import org.apache.solr.benchmarks.ControlledExecutor;
+import org.apache.solr.benchmarks.beans.TaskByClass;
+import org.apache.solr.benchmarks.prometheus.PrometheusExportManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A task similar to the simple TaskType.command but with more control (concurrent/retries/rate/duration etc)
+ */
+public class UrlCommandTask extends AbstractTask<UrlCommandTask.ExecutionResult> {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private final List<String> urls;
+  private final int maxRetry; //max retry on non zero exit code, no retry on exception
+  private final int RETRY_DELAY = 10000; //10 secs delay
+  private final String taskName;
+  private final UrlResponseListener listener;
+  private final AtomicInteger runCounter = new AtomicInteger();
+
+  public static UrlCommandTask buildTask(TaskByClass taskSpec, List<String> urls) {
+    if (taskSpec.durationSecs == null && taskSpec.rpm == null) { //then we assume only run per urls once only
+      return new UrlCommandTask(taskSpec, urls, (long)urls.size());
+    } else {
+      return new UrlCommandTask(taskSpec, urls, null);
```

Comment:
If `taskSpec.rpm` is non-null but `taskSpec.durationSecs` is null, wouldn't this throw an exception in the superclass constructor? And `durationSecs` isn't required if you want an infinite task, or am I wrong there?

Reply:
>If taskSpec.rpm is non-null but taskSpec.durationSecs is null, wouldn't this throw an exception in the superclass constructor?

Your assessment is correct! 😊 . And it is expected to throw such exception from super ctor. The ctor `if (taskSpec.durationSecs == null && taskSpec.rpm == null)` check here is not a validation, but rather a special case for `UrlCommandTask`. 

The validation is still done in the super ctor, which `taskSpec.rpm is non-null but taskSpec.durationSecs is null` is considered invalid config as the task would not be able to stop if `rpm` is defined (hence expecting it to repeat), but has no execution limit (no duration or max execution count)

>And durationSecs isn't required if you want an infinite task, or am I wrong there?

Either `durationSecs` or `maxExecution` is required, and the previous notion of `finite/infinite` tasks that I used before was a bit misleading. All the tasks should be finite in some sense - that it should know when to stop, whether by duration or execution count. 

We might allow some "infinite" tasks if we add a "run until" support - for example a background task that can just run until another task ends, this will be helpful, but i had already 2 attempts to do it and failed to come up with something satisfactory yet simple enough 😭